File: school/processess/homework4/water.cpp
#include "pthread.hpp"

#include <cstdio>
#include <vector>

/*
 * This program is developed using gcc 3. You may have problems compileing it
 * with older versions of gcc since they do not handle namespaces, c-headers
 * and templates according to the standard. I have compiled the code with gcc
 * 3.3 compilers on both linux and solaris 5.9 (my.nada.kth.se) and it works
 * fine (at least to compile).
 * To compile on solaris:
 * g++ -Wall water.cpp -lpthread -lposix4
 * To compile on linux:
 * g++ -Wall water.cpp -lpthread
*/

static const int FREE = 0;
static const int READY = 1;
static const int CONSUMED = 2;

class WaterCreator {	
private:
  int *m_hready; // Status of the H atoms
  cMutex m_mutex;
  cCondVar m_cond;
  int m_hcnt; // Number of arrived H atoms
  int m_hdone; // index of first unconsumed H atom
  int m_ncreated; // Number of created molecules.
  
  WaterCreator(const WaterCreator &c);
public:
  WaterCreator(const int no) : m_hready(new int[no*2]), m_mutex(), m_cond(), m_hcnt(0), m_hdone(0), m_ncreated(0) {
  }
  ~WaterCreator() {
    delete [] m_hready;
  }

  void oReady() {
    m_mutex.wait();    
    int hindex = m_hdone; // Pick two H-atoms to wait for.
    m_hdone += 2;
    printf("An O is waiting for H: %d and %d\n", hindex, hindex+1);

    m_cond.signalAll();
    
    m_mutex.unlock();
    m_cond.wait(m_mutex, std::equal_to<int>(), m_hready[hindex], READY);
    m_mutex.unlock();
    m_cond.wait(m_mutex, std::equal_to<int>(), m_hready[hindex+1], READY);
    /* Only one thread can make water since the mutex is taken here. */
    makeWater(hindex);
    m_mutex.unlock();
  }

  void hReady() {
    m_mutex.wait();
    int hindex = m_hcnt; // Get who I am.
    m_hcnt++;
    m_hready[hindex] = READY;
    m_cond.signalAll();
    
    m_mutex.unlock();
    m_cond.wait(m_mutex, std::equal_to<int>(), m_hready[hindex], CONSUMED);
    m_mutex.unlock();
  }
  void makeWater(int hindex) {
    m_hready[hindex] = CONSUMED;
    m_hready[hindex+1] = CONSUMED;
    m_cond.signalAll();
    printf("Made water molecule number: %d\n", ++m_ncreated);
  }
};

class O : public aPThread {
private:
  WaterCreator *m_creator;
  int m_id;
public:
  O(WaterCreator *w, int id) : m_creator(w), m_id(id) {
  }
  ~O() {
  }
  void run() {
    struct timespec sleeptime = {cRandomGenerator::get_instance().get_rand() % 4, cRandomGenerator::get_instance().get_rand()};
    nanosleep(&sleeptime, NULL);
    printf("O atom %d is ready\n", m_id);
    m_creator->oReady();
    printf("O atom %d was consumed.\n", m_id);
  }
};

class H : public aPThread {
private:
  WaterCreator *m_creator;
  int m_id;
public:
  H(WaterCreator *w, int id) : m_creator(w), m_id(id) {
  }
  ~H() {
  }
  void run() {
    struct timespec sleeptime = {cRandomGenerator::get_instance().get_rand() % 4, cRandomGenerator::get_instance().get_rand()};
    nanosleep(&sleeptime, NULL);
    printf("H atom %d is ready\n", m_id);
    m_creator->hReady();
    printf("H %d atom was consumed.\n", m_id);
  }
};

int main(int argc, char *argv[]) {
  if (argc != 2) {
    printf("Usage: %s number_of_O_atoms\n", argv[0]);
    return 0;
  }

  /* Init random generator. */
  cRandomGenerator::get_instance().spawn();

  int nO = atoi(argv[1]);
  WaterCreator creator(nO);
  std::vector<O*> o;
  std::vector<H*> h;

  for (int i = 0; i < nO; ++i) {
    o.push_back(new O(&creator, i));
    o.back()->spawn();
  }
  for (int i = 0; i < nO*2; ++i) {
    h.push_back(new H(&creator, i));
    h.back()->spawn();
  }
  
  for (std::vector<O*>::iterator itr = o.begin(); itr != o.end(); ++itr) {
    (*itr)->wait();
    delete *itr;
  }
  for (std::vector<H*>::iterator itr = h.begin(); itr != h.end(); ++itr) {
    (*itr)->wait();
    delete *itr;
  }

  /* Kill the random generator thread. */
  cRandomGenerator::get_instance().kill();
  cRandomGenerator::get_instance().wait();
  
  puts("All H and O atoms consumed.");
  
  return 0;
}

File: school/processess/homework4/Water/Water.java
/*
 * Compile with:
 * javac Water.java
 * Run with:
 * java Water
 * */


class WaterCreator {
	public static final int FREE = 0;
	public static final int READY = 1;
	public static final int CONSUMED = 2;
	
	private int m_hready[];
	private int m_hcnt; // Number of arrived H atoms
	private int m_hdone; // index of first unconsumed H atom
	private int m_ncreated;
	
	public WaterCreator(int no) {
		m_hready = new int[no*2];
		m_ncreated = m_hdone = m_hcnt = 0;
	}
	public synchronized void oReady() {		
		int hindex = m_hdone; // Pick two H-atoms to wait for.
		m_hdone += 2;
		System.out.println("An O is waiting for H: " + hindex + " and " + (hindex+1));
		notifyAll();
		
		while (m_hready[hindex] == FREE) {
			try {
				wait();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		while (m_hready[hindex+1] == FREE) {
			try {
				wait();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		makeWater(hindex);
	}
	public synchronized void hReady() {
		int hindex = m_hcnt; // Get who I am.
		m_hcnt++;
		m_hready[hindex] = READY;
		notifyAll();
		
		while (m_hready[hindex] == READY) {
			try {
				wait();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
	private synchronized void makeWater(int hindex) {
		m_hready[hindex] = CONSUMED;
		m_hready[hindex+1] = CONSUMED;
		notifyAll();
		++m_ncreated;
		System.out.println("Made water molecule number: " + m_ncreated);
	}
}

class O extends Thread {
	private WaterCreator m_creator;
	public O(WaterCreator w) {
		m_creator = w;
	}
	public void run() {
		try {
			sleep((int)(Math.random()*10));
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		System.out.println("An O atom is ready");
		m_creator.oReady();
		System.out.println("An O atom was consumed.");
	}
}

class H extends Thread {
	private WaterCreator m_creator;
	public H(WaterCreator w) {
		m_creator = w;
	}
	public void run() {
		try {
			sleep((int)(Math.random()*10000));
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		System.out.println("A H atom is ready");
		m_creator.hReady();
		System.out.println("A H atom was consumed.");
	}
}

public class Water {
	public static void main(String[] args) {
		final int nO = 200;
		WaterCreator creator = new WaterCreator(nO);
		O o[] = new O[nO];
		H h[] = new H[nO*2];
		for (int i = 0; i < nO; ++i) {
			o[i] = new O(creator);
			o[i].start();
		}
		for (int i = 0; i < nO*2; ++i) {
			h[i] = new H(creator);
			h[i].start();
		}
		
		for (int i = 0; i < nO; ++i) {
			try {
				o[i].join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		for (int i = 0; i < nO*2; ++i) {
			try {
				h[i].join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		System.out.println("All H and O atoms consumed.");
	}
}
